### PR TITLE
feat(filterFilter): allow overwriting the special `$` property name

### LINKED ISSUE #13313: Unable to watch the changes of the property '$' of an object which is used by filterFilter.
Unable to watch the changes of the property '$' of an object which is used by filterFilter.

```
//object to search 
$scope.search = {'$': 'something...'}

//watch the object
$scope.$watch('search',
  function (newValue, oldValue) {
    //do action 
    results = filter(list, search);
    ...
  }, true);
```
The "$" in the function filterFilter represents any property, however it represents built-in variable in the function equals(watch). 
How to resolve the conflict between them?

filterFilter: https://github.com/angular/angular.js/blob/master/src/ng/filter/filter.js#L235
equals(watch): https://github.com/angular/angular.js/blob/master/src/Angular.js#L992
Smart-Table: https://github.com/lorenzofox3/Smart-Table/blob/master/src/stTable.js#L122

### PR DESCRIPTION
**Pull request by gkalpak**
Previously, the special property name that would match against any
property was hard-coded to `$`.
With this commit, the user can specify an arbitrary property name,
by passing a 4th argument to `filterFilter()`. E.g.:

``` js
var items = [{foo: 'bar'}, {baz: 'qux'}];
var expr  =  {'%': 'bar'};

console.log(filterFilter(items, expr, null, '%'));   // [{foo: 'bar'}]
```

Fixes #13313

### CODE REVIEW

File: src/ng/filter/filter.js [hunk 1]
```diff
@@ -59,6 +60,9 @@
  *     Primitive values are converted to strings. Objects are not compared against primitives,
  *     unless they have a custom `toString` method (e.g. `Date` objects).
  *
+ * @param {string=} specialKey The special property name that matches against any property.
```
Comment (petebacondarwin) [inline src/ng/filter/filter.js]: Can we call this parameter `anyPropertyKey`?

**Comment by petebacondarwin** [comment]
Other than the param name, this all looks good to me. But I think we should wait for 1.5.0 to land now.

**Comment by gkalpak** [comment]
Updated the param name. I'm fine with waiting for 1.5.0 (since this is just an enhancement anyway).
I'm not sure what milestone to set. `1.5.x` ?

**Comment by Narretz** [comment]
Since this not a BC, should we add it to 1.5?

**Comment by gkalpak** [comment]
> But I think we should wait for 1.5.0 to land now.

Maybe I miunderstood @petebacondarwin's comment. He probably meant to merge this in a subsequent `1.5.x` release. He would know best what he meant though :stuck_out_tongue: 

Since this is neither a BC nor a fix, I'm fine either way.

**Comment by petebacondarwin** [comment]
I think I meant that we were about about release 1.5.0 so we should just put it in a 1.5.x release

**Pull request merged**